drivers/soundwire/intel_init.c

```diff
@@ -214,35 +254,76 @@ static struct sdw_intel_ctx
 	count = res->count;
 	dev_dbg(&adev->dev, "Creating %d SDW Link devices\n", count);
 
-	ctx = devm_kzalloc(&adev->dev, sizeof(*ctx), GFP_KERNEL);
+	/*
+	 * we need to alloc/free memory manually and can't use devm: this routine
+	 * may be called from a workqueue, and not from the parent .probe.
+	 * If devm_ was used, the memory might never be freed on errors.
+	 */
+	ctx = kzalloc(sizeof(*ctx), GFP_KERNEL);
 	if (!ctx)
 		return NULL;
 
 	ctx->count = count;
-	ctx->links = devm_kcalloc(&adev->dev, ctx->count,
-				  sizeof(*ctx->links), GFP_KERNEL);
-	if (!ctx->links)
+	ctx->ldev = kcalloc(ctx->count, sizeof(**ctx->ldev), GFP_KERNEL);
```

Comment:
this doesn't look right. `ldev` is a pointer-to-pointer, i.e. it can be pointing to an array of pointers? So what should this code be allocating - pointers or objects? Looks like it should be pointers, so this is "just" overallocating RAM, it should be just `sizeof(*ctx->ldev)`

Reply:
Good catch @lyakh !

Reply:
indeed, thanks a lot @lyakh. Fixed now and added comments on how the memory allocation works